### ISSUE TITLE
Stagemonitor no longer works under Wildfly 10 (SNAPSHOT)

### ISSUE THREAD

**Issue by domtoupin** [stagemonitor/stagemonitor#190]
While investigating issue #189 I came across the following error when starting stagemonitor under Wildfly 10:
```java
ERROR [org.stagemonitor.core.Stagemonitor] (MSC service thread 1-4) Could not install class file transformer: java.lang.IllegalStateException: Could not install class file transformer
	at net.bytebuddy.agent.builder.AgentBuilder$InstallationStrategy$Default$1.onError(AgentBuilder.java:2411)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.installOn(AgentBuilder.java:4968)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$Delegator.installOn(AgentBuilder.java:6591)
	at org.stagemonitor.core.instrument.AgentAttacher.initByteBuddyClassFileTransformer(AgentAttacher.java:159)
	at org.stagemonitor.core.instrument.AgentAttacher.performRuntimeAttachment(AgentAttacher.java:79)
	at org.stagemonitor.core.Stagemonitor.reset(Stagemonitor.java:254)
	at org.stagemonitor.core.Stagemonitor.<clinit>(Stagemonitor.java:41)
	at org.stagemonitor.core.instrument.StagemonitorRuntimeAgentAttacherDriver.<clinit>(StagemonitorRuntimeAgentAttacherDriver.java:23)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.jboss.as.connector.deployers.ds.processors.DriverProcessor.deploy(DriverProcessor.java:65)
	at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:147)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1948)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1881)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.UnsupportedOperationException: class redefinition failed: attempted to change the class modifiers
	at sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
	at sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:144)
	at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy$Collector$ForRetransformation$Cumulative.apply(AgentBuilder.java:2932)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.installOn(AgentBuilder.java:4964)
	... 17 more
```

It seems to be related to a new version of bytebuddy as reverting to bytebuddy 1.4.7 (as in version 0.27.2) fixes the problem.

Element | Version
------ | -----
Stagemonitor | 0.28.0-SNAPSHOT
Java | "1.8.0_101"
Wildfly | 10.0.0.Final

**Comment by felixbarny** [comment]
Have you tried newer versions of Byte Buddy? Could you identify the Byte Buddy version at which the error first occurred? 

/cc @raphw

**Comment by raphw** [comment]
I introduced a new mechanism for fixing inconsistent byte code in this version. Could you try version 1.4.20?

**Comment by domtoupin** [comment]
I tried versions 1.4.21-SNAPSHOT, 1.4.20, 1.4.10 and 1.4.9; only 1.4.9 works, therefore the problem is most likely caused by something introduced in 1.4.10.

**Comment by domtoupin** [comment]
I am not sure but might this be a side-effect of this commit: https://github.com/raphw/byte-buddy/commit/1eb61090ad79ed6440f534b025ac23b1c809f779 ?

**Comment by raphw** [comment]
No, the later is just a cache optimization, it should not change the outcome. I just looked at the changes made in `1.4.10` and at your error message and I am pretty sure it is my adoption of better adjusting modifier values from a `TypePool`.

Fun fact: the modifiers found in a class file and the modifiers exposed at runtime are not equal. Therefore, Byte Buddy needs to bridge between the two. There were a few bugs prior to 1.4.10 which I fixed but it seems like I also added a new error.

Could you run the transformation with `-Dnet.bytebuddy.dump=/some/location` and send the class files? Ideally, you would run the agent with a *chunked* retransformation. This way, it becomes obvious what class causes the problem. @felixbarny Is this currently configurable in Stage Monitor?

I do however have another suspicion. I introduced the mentioned change with the stack clean-up in `1.4.11` where I do however still have some problems which I fix later. This required some API breaks which might cause the problem. @domtoupin Does your error message change at all when you run a newer version?

**Comment by domtoupin** [comment]
I modified StageMonitor code in order to enable _chunked_ retransformation and it would seems the issue is related to anonymous inner classes:

```
ERROR [org.stagemonitor.core.Stagemonitor] (MSC service thread 1-4) Could not install class file transformer: java.lang.IllegalStateException: Could not install class file transformer
	at net.bytebuddy.agent.builder.AgentBuilder$InstallationStrategy$Default$1.onError(AgentBuilder.java:2403)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.installOn(AgentBuilder.java:5283)
	at net.bytebuddy.agent.builder.AgentBuilder$Default$Delegator.installOn(AgentBuilder.java:7196)
	at org.stagemonitor.core.instrument.AgentAttacher.initByteBuddyClassFileTransformer(AgentAttacher.java:159)
	at org.stagemonitor.core.instrument.AgentAttacher.performRuntimeAttachment(AgentAttacher.java:79)
	at org.stagemonitor.core.Stagemonitor.reset(Stagemonitor.java:254)
	at org.stagemonitor.core.Stagemonitor.<clinit>(Stagemonitor.java:41)
	at org.stagemonitor.core.instrument.StagemonitorRuntimeAgentAttacherDriver.<clinit>(StagemonitorRuntimeAgentAttacherDriver.java:23)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.jboss.as.connector.deployers.ds.processors.DriverProcessor.deploy(DriverProcessor.java:65)
	at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:147)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1948)
	at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1881)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalStateException: Could not retransform at least one class: {class com.obfuscated.SomeBean$2=java.lang.UnsupportedOperationException: class redefinition failed: attempted to change the class modifiers, class com.obfuscated.SomeBean$1=java.lang.UnsupportedOperationException: class redefinition failed: attempted to change the class modifiers}
	at net.bytebuddy.agent.builder.AgentBuilder$RedefinitionStrategy$Collector$ForRetransformation$Chunked.apply(AgentBuilder.java:3271)
	at net.bytebuddy.agent.builder.AgentBuilder$Default.installOn(AgentBuilder.java:5279)
	... 17 more
```

The anonymous inner classes are instances of Guava `com.google.common.cache.CacheLoader` class.

@raphw I will send you the corresponding ".class" files by e-mail.

**Comment by felixbarny** [comment]
> Ideally, you would run the agent with a chunked retransformation. This way, it becomes obvious what class causes the problem. @felixbarny Is this currently configurable in Stage Monitor?

Not yet, but I will add it.

There's currently a lot of things going on at work and my weekends are also quite stuffed. So sorry for not being as active as usual.

**Comment by raphw** [comment]
@domtoupin As you already created a custom build: Can you update Byte Buddy to 1.4.20 and change 

```
class AutoEvictingCachingBinaryLocator extends AgentBuilder.PoolStrategy.WithTypePoolCache
```

I changed the resolution of anonymous classes in a recent version where this exact problem should be fixed.

**Comment by domtoupin** [comment]
@raphw Sorry, I completely forgot to mention that for my previous test I used ByteBuddy 1.4.20 and already changed
```
    public class AutoEvictingCachingBinaryLocator extends AgentBuilder.TypeLocator.WithTypePoolCache
```
by
```
    public class AutoEvictingCachingBinaryLocator extends AgentBuilder.PoolStrategy.WithTypePoolCache
```

and this is how I got the previous call stack.

**Comment by raphw** [comment]
Alright, please send me the files in question to rafael (dot) wth (at) gmail (dot) com and I will have a look!

**Comment by domtoupin** [comment]
I sent you the files about 45 minutes ago; have you received them?

**Comment by raphw** [comment]
No, also checked my spam folder.

**Comment by domtoupin** [comment]
I sent them a second time; hopefully they will make it this time...

**Comment by raphw** [comment]
I received them. Could you also send me the files in their untransformed form?

**Comment by raphw** [comment]
I identified the problem. It is related to how javac renders anonymous classes; it seems like they are not always implicitly final and Byte Buddy does not currently check for this. I will fix it in the next version.

**Comment by domtoupin** [comment]
@raphw Great news!

@felixbarny Once the new version of byte-buddy is out, I will try it out and create a PR if everything works well.

Thank you both for everything!

**Comment by raphw** [comment]
@domtoupin Byte Buddy 1.4.21 is released and should fix this regression. Could you give it a try?

**Comment by domtoupin** [comment]
@raphw I tried Byte Buddy 1.4.21 and everything seems OK under Wildfly 10. I will open a PR to fix this in the next StageMonitor release. Thank you!

**Comment by felixbarny** [comment]
fixed by #192
